### PR TITLE
Selection1D and RangeXY callback support in plotly backend

### PR DESCRIPTION
**Pull request by jonmmease**
### Overview
This PR adds support the `Selection1D` and `RangeXY` streams in the plotly backend. Comments welcome, and this is obviously intended for consideration after the feature freeze 🙂 

I believe that it is working well, though I've only tested it manually on `Points` elements so far.

### Implementation
This implementation approach differs a bit from the Bokeh backend approach because plotly.py doesn't really have a notion of attaching a JavaScript callback to a plotly.js element.  Instead, the trace UID's are used to provide a mapping between traces and comms, and the renderer is responsible for producing all of the necessary callbacks to handle the active streams.

#### `Selection1D`
Here is an example of displaying a red point at the mean of the selected points
```python
import numpy as np
import holoviews as hv
from holoviews.streams import Stream, Selection1D, RangeXY
import param
hv.extension('plotly')

data = np.random.multivariate_normal((0, 0), [[1, 0.1], [0.1, 1]], (100,))
points = hv.Points(data).options(color='gray')

sel = Selection1D(source=points)
rangexy = RangeXY(source=points)

def mean_point(index):
    if index:
        means = data[index, :].mean(axis=0)
        pts1 = hv.Points([means]).options(size=12, color='red')
    else:
        pts1 = hv.Points([])
    
    return pts1

mean_sel = hv.DynamicMap(mean_point, kdims=[], streams=[sel])
points*mean_sel
```
selection1d

#### `RangeXY`
And here's an example of displaying a blue dot at the mean of all of the points in the `RangeXY` viewport.
```python
import numpy as np
import holoviews as hv
from holoviews.streams import Stream, Selection1D, RangeXY
import param
hv.extension('plotly')

data = np.random.multivariate_normal((0, 0), [[1, 0.1], [0.1, 1]], (100,))
points = hv.Points(data).options(color='gray')

sel = Selection1D(source=points)
rangexy = RangeXY(source=points)

def mean_point(x_range, y_range):

    if x_range and y_range:
        inds = np.logical_and(np.logical_and(x_range[0] <= data[:, 0], data[:, 0] <= x_range[1]),
                              np.logical_and(y_range[0] <= data[:, 1], data[:, 1] <= y_range[1]))
        x_mean = np.mean(data[inds, 0])
        y_mean = np.mean(data[inds, 1])
        
        print(x_mean, y_mean)
        pts2 = hv.Points([[x_mean, y_mean]]).options(size=12, color='blue')
    else:
        pts2 = hv.Points([])
    
    return pts2

mean_rangexy = hv.DynamicMap(mean_point, kdims=[], streams=[rangexy])
points*mean_rangexy
```
rangexy

#### Both `Selection1D` and `RangeXY`
Now with both streams active in the same overlay element

```python
import numpy as np
import holoviews as hv
from holoviews.streams import Stream, Selection1D, RangeXY
import param
hv.extension('plotly')

data = np.random.multivariate_normal((0, 0), [[1, 0.1], [0.1, 1]], (100,))
points = hv.Points(data).options(color='gray')

sel = Selection1D(source=points)
rangexy = RangeXY(source=points)

def mean_point(index, x_range, y_range):
    if index:
        means = data[index, :].mean(axis=0)
        pts1 = hv.Points([means]).options(size=12, color='red')
    else:
        pts1 = hv.Points([])

    if x_range and y_range:
        inds = np.logical_and(np.logical_and(x_range[0] <= data[:, 0], data[:, 0] <= x_range[1]),
                              np.logical_and(y_range[0] <= data[:, 1], data[:, 1] <= y_range[1]))
        x_mean = np.mean(data[inds, 0])
        y_mean = np.mean(data[inds, 1])
        
        print(x_mean, y_mean)
        pts2 = hv.Points([[x_mean, y_mean]]).options(size=12, color='blue')
    else:
        pts2 = hv.Points([])
    
    return pts1*pts2

mean_sel = hv.DynamicMap(mean_point, kdims=[], streams=[sel, rangexy])
points*mean_sel
```
selection1d_rangexy

And in a Layout (rather than Overlay) configuration
```python
points+mean_sel
```
selection1d_rangexy_layout

### Testing
There isn't any yet, do you have any suggestions here?  I have't really dug into the bokeh testing code to know if there's an existing approach to testing interactive callbacks.

### CODE REVIEW

**Comment by jbednar** [comment]
Cool!

**Comment by jbednar** [comment]
Is this now sufficient to support Datashader-backed Plotly-rendered HoloViews plots?

File: holoviews/plotting/plotly/element.py [hunk 1]
```diff
@@ -99,6 +101,55 @@ class ElementPlot(PlotlyPlot, GenericElementPlot):
     # Declare which styles cannot be mapped to a non-scalar dimension
     _nonvectorized_styles = []
 
+    def __init__(self, element, plot=None, **params):
+        super(ElementPlot, self).__init__(element, **params)
+        self.callbacks = self._construct_callbacks()
+        self.trace_uid = None
+
+    def _construct_callbacks(self):
```
Comment (philippjfr) [inline holoviews/plotting/plotly/element.py]: We should eventually factor this stuff out from the bokeh backend.

File: holoviews/plotting/plotly/renderer.py [hunk 2]
```diff
@@ -24,11 +24,141 @@
...
+
+    # Based on JS_CALLBACK from pyviz_comms
+    return """
+var plotdiv = document.getElementById('{plot_id}');
+var plotly = window._Plotly || window.Plotly;
+plotly.newPlot(plotdiv, {data}, {layout}, {config}).then(function() {{
+    var elem = document.getElementById("{plot_id}.loading");
+    elem.parentNode.removeChild(elem);
+
+    function unique_events(events) {{
```
Comment (philippjfr) [inline holoviews/plotting/plotly/renderer.py]: Could we just get most of this from ``pyviz_comms.JS_CALLBACK``?

Comment (jonmmease) [reply holoviews/plotting/plotly/renderer.py]: We could if we refactor `pyviz_comms.JS_CALLBACK` a little. The utility functions are all the same, but this requires the initialization of a list of comms (`JS_CALLBACK` is for just 1) and we don't want to logic to construct `data` here.

Comment (jonmmease) [reply holoviews/plotting/plotly/renderer.py]: Would you like me to try refactoring `pyviz_comms` a bit to reuse parts of `JS_CALLBACK` here?  That's fine with me, I just wasn't sure how tightly you want to link versions of `holoviews` and `pyviz_comms` (i.e. can `holoviews` require the newest version of `pyviz_comms` or does there need to be some fallback for older versions?).

**Comment by jonmmease** [comment]
@philippjfr, I'm going to try to pick this one up again.  I think I'll take a look at refactoring `pyviz_comms.JS_CALLBACK` to see how much of the JavaScript code can be taken from there.  Please let me know if you have any other thoughts/concerns on the general approach here.  Thanks!

**Comment by philippjfr** [comment]
We are currently debating how much of HoloViews we will end up rewriting on top of panel. This would basically give us comms for free by building on top of [Plotly bokeh model](https://pyviz-dev.github.io/panel/reference/panes/Plotly.html). We'd just have to extend that model to allow attaching callbacks to the rendered bokeh plot. I'd be very happy to hear/discuss that plan with you though. It is perhaps at least a little bit weird to reimplement plotly on top of bokeh but getting free comms, the ability to deploy the plots on bokeh server, binary transfers of data arrays and a unified rendering and deployment architecture is a pretty strong draw.

**Comment by jonmmease** [comment]
Yeah, I can see how that approach would be appealing.  So HoloViews' own comms would go away then? (or be the same as panel's).  And I guess it might make sense then to use panel for widgets instead of the built-in HoloViews widgets.  I'll take a look at the implementation of the Plotly bokeh model (and yes, that is a bit of an odd sentence at this point :slightly_smiling_face:).  Are there any simple reference examples of models that support callbacks?

**Comment by philippjfr** [comment]
>Are there any simple reference examples of models that support callbacks?

There are, have a look at the ``panel/models/vtk.ts`` and the way it syncs camera state. I also just had a look at the plotly.py's BaseFigureWidget, a lot of code could basically be reused in an equivalent bokeh model. There are definitely issues with that though, apart from being somewhat crazy, so a more minimal approach may be better.

**Comment by philippjfr** [comment]
@jonmmease I assume this will all be superceded by the panel based code?

**Comment by jonmmease** [comment]
Yes, I'll close this out.  The new one should be ready soon!

**Pull request closed without merging**